app/src/main/java/com/example/locavore/Fragments/MapFragment.java

```diff
@@ -1,66 +1,287 @@
+            bounds = new LatLngBounds(latLng, latLng);
+        }
+    }
+
+    private void displayLocation() {
+        if (currentLocation != null) {
+            LatLng latLng = new LatLng(currentLocation.getLatitude(), currentLocation.getLongitude());
+            CameraUpdate cameraUpdate = CameraUpdateFactory.newLatLngZoom(latLng, 10);
+            map.animateCamera(cameraUpdate);
+            getRequest("farms");
+            getRequest("farmersmarket");
+        }
+    }
+
+    public Bitmap getMarker(String request) {
+        IconGenerator iconGen = new IconGenerator(getContext());
+
+        // Define the size from the dimensions file
+        int shapeSize = this.getResources().getDimensionPixelSize(R.dimen.custom_marker_value);
+
+        Drawable shapeDrawable = request.equals("farms") ? ResourcesCompat.getDrawable(this.getResources(), R.drawable.ic_farmhouse, null) : ResourcesCompat.getDrawable(this.getResources(), R.drawable.ic_baseline_storefront_24, null);
+        iconGen.setBackground(shapeDrawable);
+
+        // Create a view container to set the size
+        View view = new View(getContext());
+        view.setLayoutParams(new ViewGroup.LayoutParams(shapeSize, shapeSize));
+        iconGen.setContentView(view);
+
+        // Create the bitmap
+        return iconGen.makeIcon();
+    }
+
+    protected void dropMarkers(List<Farm> newFarms, String request){
+        for(Farm farm : newFarms) {
+            map.addMarker(new MarkerOptions()
+                    .position(farm.getCoordinates())
+                    .title(farm.getName())
+                    .icon(BitmapDescriptorFactory.fromBitmap(getMarker(request)))
+            );
+            bounds = bounds.including(farm.getCoordinates());
+            map.setLatLngBoundsForCameraTarget(bounds);
+        }
+    }
+
+    protected void getRequest(String request) {
+        // network request: need to take off of main thread? does retrofit automatically do this?
+        YelpService yelpService = retrofit.create(YelpService.class);
+        Call<FarmSearchResult> call = yelpService.searchFarms("Bearer "+YELP_API_KEY, currentLocation.getLatitude(), currentLocation.getLongitude(),request, 50, radius);
+        call.enqueue(new Callback<FarmSearchResult>() {
+            @Override
+            public void onResponse(Call<FarmSearchResult> call, Response<FarmSearchResult> response) {
+                Log.i(TAG, "Success! " + response);
+                if(response.body() == null) {
+                    Log.e(TAG, "Error retrieving response body");
+                } else {
+                    List<Farm> newFarms = new ArrayList<>();
+                    for(Farm farm : response.body().getFarms()) {
+                        if(!existingFarm(farm, farms) && farm.getDistance() < radius) {
+                            newFarms.add(farm);
+                        }
+                    }
```

Comment:
Here may be a potential hot spot that costs too long time since you're using an O(n^2) algorithm to eliminate the duplicated farms. I have two questions here:
1. Is it possible that there are duplicated farms with same Id in the yelp response?
2. If the answer to question 1 is yes, then can we use following approach to accelerate it?
```
Set<String> newFarmIds;
for (Farm farm: response.body().getFarms()) {
  if (farm.getDistance() < radius && !farmIds.contains(farm.getId()) {
    newFarms.add(farm);
    newFarmsIds.add(farm.getId());
  }
}
```

Reply:
the farms should have unique ids-- great idea, thank you!